File: backend/app/schemas/message.py
"""Pydantic schemas for Direct Message API requests/responses."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class MessageCreate(BaseModel):
    """Schema for creating a new message."""

    recipient_id: uuid.UUID
    subject: str | None = Field(None, max_length=500)
    body: str = Field(..., min_length=1)
    parent_message_id: uuid.UUID | None = None

    @field_validator("body")
    @classmethod
    def validate_body_not_empty(cls, v: str) -> str:
        """Validate body is not empty or whitespace."""
        if not v.strip():
            raise ValueError("Message body cannot be empty")
        return v


class MessagePublic(BaseModel):
    """Schema for message API response with sender/recipient names."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    sender_id: uuid.UUID
    sender_name: str
    sender_email: str
    recipient_id: uuid.UUID
    recipient_name: str
    recipient_email: str
    subject: str | None
    body: str
    parent_message_id: uuid.UUID | None
    is_read: bool
    sent_at: datetime


class ConversationPublic(BaseModel):
    """Schema for a conversation (grouped messages with a participant)."""

    participant_id: uuid.UUID
    participant_name: str
    participant_email: str
    participant_role: str
    last_message_preview: str
    last_message_timestamp: datetime
    unread_count: int


class ConversationListResponse(BaseModel):
    """Schema for paginated conversation list response."""

    conversations: list[ConversationPublic]
    total: int
    limit: int
    offset: int
    has_more: bool
    total_unread: int


class MessageThreadResponse(BaseModel):
    """Schema for message thread response."""

    participant_id: uuid.UUID
    participant_name: str
    participant_email: str
    participant_role: str
    messages: list[MessagePublic]
    total_messages: int


class RecipientPublic(BaseModel):
    """Schema for an allowed recipient."""

    user_id: uuid.UUID
    name: str
    email: str
    role: str


class RecipientListResponse(BaseModel):
    """Schema for list of allowed recipients."""

    recipients: list[RecipientPublic]
    total: int


class MessageReadResponse(BaseModel):
    """Schema for message read status update response."""

    id: uuid.UUID
    is_read: bool


class UnreadMessagesCountResponse(BaseModel):
    """Schema for unread messages count response."""

    count: int

File: frontend/src/types/message.ts
/**
 * Message type definitions for Dream LMS Direct Messaging.
 * Story 6.3: Direct Messaging Between Teachers & Students
 */

/**
 * Message interface with all fields from API
 */
/**
 * System message categories for icon/color mapping
 */
export type MessageCategory =
  | "assignment_created"
  | "student_completed"
  | "feedback_received";

/**
 * Icon and color config for system message categories
 */
export const MESSAGE_CATEGORY_CONFIG: Record<
  MessageCategory,
  { icon: string; color: string; bgColor: string }
> = {
  assignment_created: {
    icon: "FileText",
    color: "text-blue-600",
    bgColor: "bg-blue-50",
  },
  student_completed: {
    icon: "CheckCircle",
    color: "text-green-600",
    bgColor: "bg-green-50",
  },
  feedback_received: {
    icon: "MessageSquare",
    color: "text-emerald-600",
    bgColor: "bg-emerald-50",
  },
};

export interface Message {
  id: string;
  sender_id: string;
  sender_name: string;
  sender_email?: string | null;
  recipient_id: string;
  recipient_name: string;
  recipient_email?: string | null;
  subject: string | null;
  body: string;
  parent_message_id: string | null;
  is_read: boolean;
  sent_at: string; // ISO 8601 datetime string
  is_system: boolean;
  context_type: string | null;
  context_id: string | null;
  message_category: MessageCategory | null;
}

/**
 * Request payload for creating a new message
 */
export interface MessageCreate {
  recipient_id: string;
  subject?: string | null;
  body: string;
  parent_message_id?: string | null;
}

/**
 * Conversation summary for conversation list
 */
export interface Conversation {
  participant_id: string;
  participant_name: string;
  participant_email?: string | null;
  participant_role: string;
  last_message_preview: string;
  last_message_timestamp: string; // ISO 8601 datetime string
  unread_count: number;
}

/**
 * Paginated conversation list response from API
 */
export interface ConversationListResponse {
  conversations: Conversation[];
  total: number;
  limit: number;
  offset: number;
  has_more: boolean;
  total_unread: number;
}

/**
 * Message thread response from API
 */
export interface MessageThreadResponse {
  participant_id: string;
  participant_name: string;
  participant_email?: string | null;
  participant_role: string;
  participant_organization_name?: string | null; // Publisher organization name
  messages: Message[];
  total_messages: number;
}

/**
 * Allowed recipient for messaging
 */
export interface Recipient {
  user_id: string;
  name: string;
  email?: string | null;
  role: string;
  organization_name?: string | null; // Publisher organization name (from DCS)
}

/**
 * Recipient list response from API
 */
export interface RecipientListResponse {
  recipients: Recipient[];
  total: number;
}

/**
 * Message read status update response
 */
export interface MessageReadResponse {
  id: string;
  is_read: boolean;
}

/**
 * Unread messages count response
 */
export interface UnreadMessagesCountResponse {
  count: number;
}

/**
 * Query parameters for fetching conversations
 */
export interface ConversationQueryParams {
  limit?: number;
  offset?: number;
}
